posts/2022-03-06-typechecking-pt1.markdown

```diff
@@ -0,0 +1,210 @@
+complicated... 
+
+#### Unifying our new constraint 
+
+Let's look at the constraint we've generated during elaboration.
+
+![Unifying function application](/images/typecheck-1-unify-application.png)
+
+As we briefly mentioned earlier, we unify values to learn things about them. Because `unknown-1` is a unification variable, it represents an unknown thing that we are trying to find out more about.
+Unifying it with `Boolean` tells us `unknown-1` is actually `Boolean`, and therefore we can substitute any instance of `unknown-1` for `Boolean`.
+
+#### Apply the substitution
+
+We apply a substitution by changing all instances of one thing in a type for another. Applying a substitution to the result of elaboration gives us our final type,
+`Boolean`
+
+![Substituting function application](/images/typecheck-1-substitute-application.png)
+
+> __Substitution__
+>
+> Assuming we had no type errors in the previous step, then
+>  subsitution is the process of taking all the things we have learned and
+>  applying them to both to expression and to the other substitutions.
+> 
+> For example, if we have learned __"`a` is an `Integer`"__ then we can change any `a`
+>  values in our expression to `Integer`, and also change the substitution __"`c`
+> is a pair of `a` and `b`"__ into __"`c` is a pair of `Integer` and `b`"__.
+
+
+
+### A bigger example
+
+To see Hindley Milner typechecking in action, here is a slightly bigger
+example. Here we'll see how we can get full type inference without a type
+annotation in sight.
+
+These diagrams are getting a bit out of hand now, so to be clear, our
+expression is `\a -> if a then 1 else 2`.
+
+![Elaborating a big lambda](/images/typecheck-1-elaborate-lambda-2.png)
+
+We start by elaborating the function argument, `a`. We have no idea what it is,
+so we create a unification variable, `unknown-1`, and use that instead.
+
+Then we move onto the body of the lambda, an if expression. The predicate
+expression is `a`. Fortunately, we already have the type for `a` in the type
+environment - so we can grab it. We don't know what `unknown-1` is yet, but we
+have another piece of information on our side - the predicate for an `if` must
+be a `Boolean`. We express this by creating a new constraint `unknown-1 ==
+Boolean`.
+
+The `then` and `else` expressions are both literals, so we can infer those
+quite simply, making `Integer` the return type of the `if` expression. There is
+also another constraint here, that both branches should have the same type -
+`Integer == Integer`.
+
+#### Unifying our new constraints
+
+We have two contraints from elaboration.
+
+First `Integer == Integer`, which teaches us nothing (but does type check).
+
+![Unifying big lambda](/images/typecheck-1-unify-lambda-2.png)
+
+Secondly `unknown-1 == Boolean` gives us a substitution.
+
+#### Apply the substitution
+
+![Substituting big lambda](/images/typecheck-1-substitute-lambda-2.png)
+
+
```

Comment:
Maybe sum up saying what the value was and what the final type was and sum up everything.